### ISSUE TITLE
Reconsider changefeed error behavior when queue gets full

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#6109]
Currently, our behavior is as follows (from the documentation):

> The server will buffer up to `changefeedQueueSize` elements (default 100,000). If the buffer limit is hit, early changes will be discarded, and the client will receive an object of the form `{error: "Changefeed cache over array size limit, skipped X elements."}` where `X` is the number of elements skipped.

The problem is that most people are not going to handle `error` documents in their code since they are rare, especially in testing.

I propose that we change the behavior of changefeeds to throw an exception if this happens instead.
It's slightly worse because it will cause the changefeed to be terminated, rather than just skipping changes but staying alive. However I doubt that a changefeed with skipped notifications is of much use anyway.

This reduces the error cases that users have to think about and handle.

This is a ReQL proposal, but for the case that we get a quick consensus, I'd like to still change this for 2.4.

@mlucy, @VeXocide, @larkost, @deontologician thoughts?

**Comment by deontologician** [comment]
Seems reasonable to me.

**Comment by VeXocide** [comment]
I agree, we're breaking the guarantee that a message will be delivered and it's not unreasonable to expect people not to handle `error` documents.

**Comment by marshall007** [comment]
We've run into cases where changefeed notifications get "silently" dropped, leading to bugs that were hard to track down, so 👍. In fact, even though we tracked down the cause, I still had no idea an `error` document was even emitted in this case until now.

**Comment by danielmewes** [comment]
> I still had no idea an error document was even emitted in this case until now.

Exactly what I'm worried about. Thanks for the additional data point.
I think even our example codes for how to handle changefeeds don't deal with `error` documents.

**Comment by 20zinnm** [comment]
My understanding so far is that it's well documented that change-feeds only store a certain number of changes. Therefor, it's not an error that some were omitted, is it? Honestly, it's generous that the database even includes the number omitted.

**Comment by mlucy** [comment]
+1, I think by default we should abort the changefeed in this case.

**Comment by danielmewes** [comment]
@mlucy Do you think we should add an option to recover the old behavior?
I'm not sure if that's necessary.

**Comment by grandquista** [comment]
As a data point. I have a personal project where I was planning for dropped changes. It's not feasible/critical to see all changes. The feed is already designed to resume on the client side at any point. So I will be adding logic to catch the thrown error and restart the feed if this goes through.

**Comment by larkost** [comment]
This was the behavior that I expected while starting to write tests, so I am generically in favor of it. @danielmewes, would this be a complete change, or would there be an opt-arg allowing users to specifically opt-in to getting the error as a document?

And the second question is: are we planning to just change the clients (relatively easy)? Or is this a sever-side change as well?

**Comment by danielmewes** [comment]
@larkost This would definitely be a server-side change (and a pretty easy one I think). No change on the client side will be required.

I'd not add an opt-arg. I can't see a use case in which it is better to have a changefeed with a whole in the middle, compared to just restarting the feed from scratch in case the buffer overflows.
